packages-content-model/roosterjs-content-model-plugins/lib/tableEdit/editors/features/TableInserter.ts

```diff
@@ -0,0 +1,191 @@
+                    : tableRect.left - (INSERTER_SIDE_LENGTH - 1 + 2 * INSERTER_BORDER_SIZE)
+            }px`;
+            div.style.top = `${tdRect.bottom - 8}px`;
+            (div.firstChild as HTMLElement).style.width = `${tableRect.right - tableRect.left}px`;
+        } else {
+            div.style.left = `${isRTL ? tdRect.left - 8 : tdRect.right - 8}px`;
+            // tableRect.top is used because the Inserter is always intended to be on top
+            div.style.top = `${
+                tableRect.top - (INSERTER_SIDE_LENGTH - 1 + 2 * INSERTER_BORDER_SIZE)
+            }px`;
+            (div.firstChild as HTMLElement).style.height = `${tableRect.bottom - tableRect.top}px`;
+        }
+
+        (anchorContainer || document.body).appendChild(div);
+
+        const handler = new TableInsertHandler(
+            div,
+            td,
+            table,
+            isHorizontal,
+            editor,
+            onInsert,
+            getOnMouseOut
+        );
+
+        return { div, featureHandler: handler, node: td };
+    }
+
+    return null;
+}
+
+class TableInsertHandler implements Disposable {
+    private onMouseOutEvent: null | ((ev: MouseEvent) => void);
+    constructor(
+        private div: HTMLDivElement,
+        private td: HTMLTableCellElement,
+        private table: HTMLTableElement,
+        private isHorizontal: boolean,
+        private editor: IStandaloneEditor,
+        private onInsert: () => void,
+        getOnMouseOut: (feature: HTMLElement) => (ev: MouseEvent) => void
+    ) {
+        this.div.addEventListener('click', this.insertTd);
+        this.onMouseOutEvent = getOnMouseOut(div);
+        this.div.addEventListener('mouseout', this.onMouseOutEvent);
+    }
+
+    dispose() {
+        this.div.removeEventListener('click', this.insertTd);
+
+        if (this.onMouseOutEvent) {
+            this.div.removeEventListener('mouseout', this.onMouseOutEvent);
+        }
+
+        this.onMouseOutEvent = null;
+    }
+
+    private insertTd = () => {
+        // Get cell coordinates
+        const columnIndex = this.td.cellIndex;
+        const row =
+            this.td.parentElement instanceof HTMLTableRowElement
+                ? this.td.parentElement
+                : undefined;
+        const rowIndex = row?.rowIndex;
+
+        if (rowIndex == undefined) {
+            return;
+        }
+
+        // Take snapshot before selection change
+        this.editor.takeSnapshot();
+
+        // Select cell to make insertion
+        this.editor.setDOMSelection({
+            type: 'table',
+            firstColumn: columnIndex,
+            firstRow: rowIndex,
+            lastColumn: columnIndex,
+            lastRow: rowIndex,
+            table: this.table,
+        });
+
+        // Insert row or column, skip snapshot as it was already taken
+        editTable(
+            this.editor,
+            this.isHorizontal ? 'insertBelow' : 'insertRight',
+            true /* skipSnapshot */
```

Comment:
If the snapshot is made inside editTable it will be a wrong.

Reply:
Can you add more detail here? What will be wrong?

Reply:
I mentioned in a previous comment:

The workflow is:
1. Click the table inserter
2. Internally the inserter selects a cell in the table and calls editTable after
3. editTable gets the table with a selected cell and does the operation

The problem is that editTable will create a snapshot with the selected cell, so between 1 and 2, a pre-selection snapshot is created and editTable is made to not create another one.

Reply:
After some thinking, I think we can do the format and selection like this:

```ts
// Replace from line 120
this.editor.formatContentModel(model => {
  const [tableModel, path] = getFirstSelectedTable(model);

  if (isHorizontal) {
     insertTableRow(tableModel, 'insertBelow');
 } else {
     insertTableColumn(tableModel, 'insertBelow');
} 

  // Table normalization and ensure empty paragraph here, you can extract from line 107 to line 123 of editTable.ts to be a new public function and call it here

  return true;
}, {
  selectionOverride: { // This will override the original selection to create a new Content Model, and no additional snapshot added
    type: 'table',
    table: table,
    firstRow: rowIndex,
    lastRow: rowIndex,
    firstColumn: columnIndex,
    lastColumn: columnIndex  
  },
  changeSource: ChangeSource.Format
});
```

So we can still change selection in model, no addition undo snapshot added, and no need to add new parameter to editTable.ts

Reply:
To make it easier, I created a draft PR here https://github.com/microsoft/roosterjs/pull/2399 to extract a common API `formatTableWithContentModel`, and let existing `editTable` API use it. You can also call it from your code here and pass in the overridden table selection.